source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,215 @@
+/*******************************************************************************
+
+    Contains MerkleTree that makes binary tree of transactions.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.MerkleTree;
+
+import agora.common.Data;
+import agora.common.Hash;
+import agora.common.Transaction;
+
+/// Ditto
+public struct MerkleTree
+{
+    /// The root of `MerkleTree`
+    public MerkleNode* root;
+
+    /// Array of `Transaction`
+    /// All transactions in the `MerkleTree` are stored here,
+    /// and refer to its memory everywhere.
+    public Transaction[] txs;
+
+    /// Array of MerkleNode,
+    /// All `MerkleNode` in the `MerkleTree` are stored here,
+    /// and refer to its memory everywhere.
+    public MerkleNode[] nodes;
+
+    /***************************************************************************
+
+        Implements hashing support
+
+        Params:
+            dg = hashing function
+
+    ***************************************************************************/
+
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        if (root !is null)
+            hashPart(*root, dg);
+    }
+
+    /***************************************************************************
+
+        Calculate the hash value of the `MerkleTree`
+
+        Return:
+            The `Hash` representing this instance
+
+    ***************************************************************************/
+
+    public Hash getHash()
+        nothrow @nogc @trusted
+    {
+        if (this.txs.length == 0)
+            return Hash.init;
+        else if (this.root is null)
+            return Hash.init;
+        else
+            return hashFull(this);
+    }
+}
+
+/// The `MerkleNode` has one transaction or two `MerkleNode`
+public struct MerkleNode
+{
+    /// The left branch's ptr of the tree.
+    public MerkleNode* left;
+
+    /// The right branch's ptr of the tree.
+    public MerkleNode* right;
+
+    /// The transaction's ptr
+    public Transaction* tx;
+
+    /// The hash of this `MerkleNode`
+    public Hash hash;
```

Comment:
This is never written to from within this struct. So it is assumed that the caller will set it ?